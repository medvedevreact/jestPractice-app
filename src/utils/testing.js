import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component, options = {}) => {
  const Wrapper = ({ children }) => (
    <MemoryRouter {...options}>{children}</MemoryRouter>
  );
  return render(component, { wrapper: Wrapper });
};
