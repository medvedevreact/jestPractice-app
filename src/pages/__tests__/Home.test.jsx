import { renderWithRouter } from '../../utils/testing';
import * as api from '../../api';
import { waitForElementToBeRemoved, screen } from '@testing-library/react';
import { Home } from '../Home';
import userEvent from '@testing-library/user-event';

const apiSpy = jest.spyOn(api, 'getAllCategories');

describe('Category', () => {
  test('render', async () => {
    apiSpy.mockResolvedValueOnce({
      categories: [
        {
          idCategory: '1',
          strCategory: 'Beef',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/beef.png',
          strCategoryDescription:
            'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.[2]',
        },
        {
          idCategory: '2',
          strCategory: 'Chicken',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/chicken.png',
          strCategoryDescription:
            'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
        },
      ],
    });

    renderWithRouter(<Home />);
    const preloader = screen.getByRole('progressbar');
    expect(preloader).toBeInTheDocument();
    await waitForElementToBeRemoved(preloader);
    expect(screen.getAllByRole('article')).toMatchSnapshot();
  });
  test('renders search bar and filters categories correctly', async () => {
    apiSpy.mockResolvedValue({
      categories: [
        {
          idCategory: '1',
          strCategory: 'Beef',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/beef.png',
          strCategoryDescription:
            'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.[2]',
        },
        {
          idCategory: '2',
          strCategory: 'Chicken',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/chicken.png',
          strCategoryDescription:
            'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
        },
      ],
    });

    renderWithRouter(<Home />);

    const preloader = screen.getByRole('progressbar');
    expect(preloader).toBeInTheDocument();
    await waitForElementToBeRemoved(preloader);

    expect(screen.getAllByRole('article')).toHaveLength(2);

    const searchInput = screen.getByPlaceholderText(/search/i);
    await userEvent.type(searchInput, 'Beef');
    await userEvent.click(screen.getByRole('button'));

    expect(screen.getAllByRole('article')).toHaveLength(1);
    expect(screen.getByText('Beef')).toBeInTheDocument();
  });
  test('should render Home with search', async () => {
    apiSpy.mockResolvedValueOnce({
      categories: [
        {
          idCategory: '1',
          strCategory: 'Beef',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/beef.png',
          strCategoryDescription:
            'Beef is the culinary name for meat from cattle, particularly skeletal muscle. Humans have been eating beef since prehistoric times.[1] Beef is a source of high-quality protein and essential nutrients.[2]',
        },
        {
          idCategory: '2',
          strCategory: 'Chicken',
          strCategoryThumb:
            'https://www.themealdb.com/images/category/chicken.png',
          strCategoryDescription:
            'Chicken is a type of domesticated fowl, a subspecies of the red junglefowl. It is one of the most common and widespread domestic animals, with a total population of more than 19 billion as of 2011.[1] Humans commonly keep chickens as a source of food (consuming both their meat and eggs) and, more rarely, as pets.',
        },
      ],
    });

    renderWithRouter(<Home />, {
      initialEntries: ['/?search=Beef'],
    });

    const preloader = screen.getByRole('progressbar');

    expect(preloader).toBeInTheDocument();

    await waitForElementToBeRemoved(preloader);

    expect(screen.getAllByRole('article')).toHaveLength(1);
  });
});
