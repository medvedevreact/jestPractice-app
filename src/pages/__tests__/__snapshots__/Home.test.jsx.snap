// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category render 1`] = `
[
  <div
    class="card"
    role="article"
  >
    <div
      class="card-image"
    >
      <img
        alt="Beef"
        src="https://www.themealdb.com/images/category/beef.png"
      />
    </div>
    <div
      class="card-content"
    >
      <span
        class="card-title"
      >
        Beef
      </span>
      <p>
        Beef is the culinary name for meat from cattle, particularly
        ...
      </p>
    </div>
    <div
      class="card-action"
    >
      <a
        class="btn"
        href="/category/Beef"
      >
        Watch category
      </a>
    </div>
  </div>,
  <div
    class="card"
    role="article"
  >
    <div
      class="card-image"
    >
      <img
        alt="Chicken"
        src="https://www.themealdb.com/images/category/chicken.png"
      />
    </div>
    <div
      class="card-content"
    >
      <span
        class="card-title"
      >
        Chicken
      </span>
      <p>
        Chicken is a type of domesticated fowl, a subspecies of the 
        ...
      </p>
    </div>
    <div
      class="card-action"
    >
      <a
        class="btn"
        href="/category/Chicken"
      >
        Watch category
      </a>
    </div>
  </div>,
]
`;
